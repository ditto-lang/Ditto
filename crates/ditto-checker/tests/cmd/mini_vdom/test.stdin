module Test exports (..);

type Html(msg);

type Attr = Attr(String, String);

div = fn (
    attrs: Array(Attr), 
    children: Array(Html(msg)),
): Html(msg) -> h("div", attrs, children);

foreign h : (String, Array(Attr), Array(Html(msg))) -> Html(msg);
