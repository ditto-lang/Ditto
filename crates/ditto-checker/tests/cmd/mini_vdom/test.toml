bin.name = "ditto-checker-testbin-check-module"
args = ["mini_vdom.ditto"]
fs.sandbox = true
