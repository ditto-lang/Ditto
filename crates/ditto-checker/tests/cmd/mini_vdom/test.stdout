{
  "module_name": [
    "Test"
  ],
  "exports": {
    "types": {
      "Attr": {
        "Type": {
          "doc_comments": [],
          "doc_position": 0,
          "kind": "Type"
        }
      },
      "Html": {
        "Type": {
          "doc_comments": [],
          "doc_position": 1,
          "kind": {
            "Function": {
              "parameters": [
                {
                  "Variable": 1
                }
              ]
            }
          }
        }
      }
    },
    "constructors": {
      "Attr": {
        "doc_comments": [],
        "doc_position": 0,
        "constructor_type": {
          "type": "Function",
          "data": {
            "parameters": [
              {
                "type": "PrimConstructor",
                "data": "String"
              },
              {
                "type": "PrimConstructor",
                "data": "String"
              }
            ],
            "return_type": {
              "type": "Constructor",
              "data": {
                "constructor_kind": "Type",
                "canonical_value": {
                  "module_name": [
                    null,
                    [
                      "Test"
                    ]
                  ],
                  "value": "Attr"
                },
                "source_value": {
                  "module_name": null,
                  "value": "Attr"
                }
              }
            }
          }
        },
        "return_type_name": "Attr"
      }
    },
    "values": {
      "div": {
        "doc_comments": [],
        "doc_position": 0,
        "value_type": {
          "type": "Function",
          "data": {
            "parameters": [
              {
                "type": "Call",
                "data": {
                  "function": {
                    "type": "PrimConstructor",
                    "data": "Array"
                  },
                  "arguments": [
                    {
                      "type": "Constructor",
                      "data": {
                        "constructor_kind": "Type",
                        "canonical_value": {
                          "module_name": [
                            null,
                            [
                              "Test"
                            ]
                          ],
                          "value": "Attr"
                        },
                        "source_value": {
                          "module_name": null,
                          "value": "Attr"
                        }
                      }
                    }
                  ]
                }
              },
              {
                "type": "Call",
                "data": {
                  "function": {
                    "type": "PrimConstructor",
                    "data": "Array"
                  },
                  "arguments": [
                    {
                      "type": "Call",
                      "data": {
                        "function": {
                          "type": "Constructor",
                          "data": {
                            "constructor_kind": {
                              "Function": {
                                "parameters": [
                                  {
                                    "Variable": 1
                                  }
                                ]
                              }
                            },
                            "canonical_value": {
                              "module_name": [
                                null,
                                [
                                  "Test"
                                ]
                              ],
                              "value": "Html"
                            },
                            "source_value": {
                              "module_name": null,
                              "value": "Html"
                            }
                          }
                        },
                        "arguments": [
                          {
                            "type": "Variable",
                            "data": {
                              "variable_kind": {
                                "Variable": 1
                              },
                              "var": 0,
                              "source_name": "msg"
                            }
                          }
                        ]
                      }
                    }
                  ]
                }
              }
            ],
            "return_type": {
              "type": "Call",
              "data": {
                "function": {
                  "type": "Constructor",
                  "data": {
                    "constructor_kind": {
                      "Function": {
                        "parameters": [
                          {
                            "Variable": 1
                          }
                        ]
                      }
                    },
                    "canonical_value": {
                      "module_name": [
                        null,
                        [
                          "Test"
                        ]
                      ],
                      "value": "Html"
                    },
                    "source_value": {
                      "module_name": null,
                      "value": "Html"
                    }
                  }
                },
                "arguments": [
                  {
                    "type": "Variable",
                    "data": {
                      "variable_kind": {
                        "Variable": 1
                      },
                      "var": 0,
                      "source_name": "msg"
                    }
                  }
                ]
              }
            }
          }
        }
      }
    }
  },
  "types": {
    "Attr": {
      "Type": {
        "doc_comments": [],
        "type_name_span": {
          "start_offset": 49,
          "end_offset": 53
        },
        "kind": "Type"
      }
    },
    "Html": {
      "Type": {
        "doc_comments": [],
        "type_name_span": {
          "start_offset": 32,
          "end_offset": 36
        },
        "kind": {
          "Function": {
            "parameters": [
              {
                "Variable": 1
              }
            ]
          }
        }
      }
    }
  },
  "constructors": {
    "Attr": {
      "doc_comments": [],
      "doc_position": 0,
      "constructor_name_span": {
        "start_offset": 56,
        "end_offset": 60
      },
      "fields": [
        {
          "type": "PrimConstructor",
          "data": "String"
        },
        {
          "type": "PrimConstructor",
          "data": "String"
        }
      ],
      "return_type": {
        "type": "Constructor",
        "data": {
          "constructor_kind": "Type",
          "canonical_value": {
            "module_name": [
              null,
              [
                "Test"
              ]
            ],
            "value": "Attr"
          },
          "source_value": {
            "module_name": null,
            "value": "Attr"
          }
        }
      },
      "return_type_name": "Attr"
    }
  },
  "values": {
    "div": {
      "doc_comments": [],
      "name_span": {
        "start_offset": 79,
        "end_offset": 82
      },
      "expression": {
        "expression": "Function",
        "data": {
          "span": {
            "start_offset": 85,
            "end_offset": 188
          },
          "function_type": {
            "type": "Function",
            "data": {
              "parameters": [
                {
                  "type": "Call",
                  "data": {
                    "function": {
                      "type": "PrimConstructor",
                      "data": "Array"
                    },
                    "arguments": [
                      {
                        "type": "Constructor",
                        "data": {
                          "constructor_kind": "Type",
                          "canonical_value": {
                            "module_name": [
                              null,
                              [
                                "Test"
                              ]
                            ],
                            "value": "Attr"
                          },
                          "source_value": {
                            "module_name": null,
                            "value": "Attr"
                          }
                        }
                      }
                    ]
                  }
                },
                {
                  "type": "Call",
                  "data": {
                    "function": {
                      "type": "PrimConstructor",
                      "data": "Array"
                    },
                    "arguments": [
                      {
                        "type": "Call",
                        "data": {
                          "function": {
                            "type": "Constructor",
                            "data": {
                              "constructor_kind": {
                                "Function": {
                                  "parameters": [
                                    {
                                      "Variable": 1
                                    }
                                  ]
                                }
                              },
                              "canonical_value": {
                                "module_name": [
                                  null,
                                  [
                                    "Test"
                                  ]
                                ],
                                "value": "Html"
                              },
                              "source_value": {
                                "module_name": null,
                                "value": "Html"
                              }
                            }
                          },
                          "arguments": [
                            {
                              "type": "Variable",
                              "data": {
                                "variable_kind": {
                                  "Variable": 1
                                },
                                "var": 0,
                                "source_name": "msg"
                              }
                            }
                          ]
                        }
                      }
                    ]
                  }
                }
              ],
              "return_type": {
                "type": "Call",
                "data": {
                  "function": {
                    "type": "Constructor",
                    "data": {
                      "constructor_kind": {
                        "Function": {
                          "parameters": [
                            {
                              "Variable": 1
                            }
                          ]
                        }
                      },
                      "canonical_value": {
                        "module_name": [
                          null,
                          [
                            "Test"
                          ]
                        ],
                        "value": "Html"
                      },
                      "source_value": {
                        "module_name": null,
                        "value": "Html"
                      }
                    }
                  },
                  "arguments": [
                    {
                      "type": "Variable",
                      "data": {
                        "variable_kind": {
                          "Variable": 1
                        },
                        "var": 0,
                        "source_name": "msg"
                      }
                    }
                  ]
                }
              }
            }
          },
          "binders": [
            [
              {
                "Variable": {
                  "span": {
                    "start_offset": 94,
                    "end_offset": 99
                  },
                  "name": "attrs"
                }
              },
              {
                "type": "Call",
                "data": {
                  "function": {
                    "type": "PrimConstructor",
                    "data": "Array"
                  },
                  "arguments": [
                    {
                      "type": "Constructor",
                      "data": {
                        "constructor_kind": "Type",
                        "canonical_value": {
                          "module_name": [
                            null,
                            [
                              "Test"
                            ]
                          ],
                          "value": "Attr"
                        },
                        "source_value": {
                          "module_name": null,
                          "value": "Attr"
                        }
                      }
                    }
                  ]
                }
              }
            ],
            [
              {
                "Variable": {
                  "span": {
                    "start_offset": 119,
                    "end_offset": 127
                  },
                  "name": "children"
                }
              },
              {
                "type": "Call",
                "data": {
                  "function": {
                    "type": "PrimConstructor",
                    "data": "Array"
                  },
                  "arguments": [
                    {
                      "type": "Call",
                      "data": {
                        "function": {
                          "type": "Constructor",
                          "data": {
                            "constructor_kind": {
                              "Function": {
                                "parameters": [
                                  {
                                    "Variable": 1
                                  }
                                ]
                              }
                            },
                            "canonical_value": {
                              "module_name": [
                                null,
                                [
                                  "Test"
                                ]
                              ],
                              "value": "Html"
                            },
                            "source_value": {
                              "module_name": null,
                              "value": "Html"
                            }
                          }
                        },
                        "arguments": [
                          {
                            "type": "Variable",
                            "data": {
                              "variable_kind": {
                                "Variable": 1
                              },
                              "var": 0,
                              "source_name": "msg"
                            }
                          }
                        ]
                      }
                    }
                  ]
                }
              }
            ]
          ],
          "body": {
            "expression": "Call",
            "data": {
              "span": {
                "start_offset": 163,
                "end_offset": 188
              },
              "call_type": {
                "type": "Call",
                "data": {
                  "function": {
                    "type": "Constructor",
                    "data": {
                      "constructor_kind": {
                        "Function": {
                          "parameters": [
                            {
                              "Variable": 1
                            }
                          ]
                        }
                      },
                      "canonical_value": {
                        "module_name": [
                          null,
                          [
                            "Test"
                          ]
                        ],
                        "value": "Html"
                      },
                      "source_value": {
                        "module_name": null,
                        "value": "Html"
                      }
                    }
                  },
                  "arguments": [
                    {
                      "type": "Variable",
                      "data": {
                        "variable_kind": {
                          "Variable": 1
                        },
                        "var": 0,
                        "source_name": "msg"
                      }
                    }
                  ]
                }
              },
              "function": {
                "expression": "ForeignVariable",
                "data": {
                  "span": {
                    "start_offset": 163,
                    "end_offset": 164
                  },
                  "variable_type": {
                    "type": "Function",
                    "data": {
                      "parameters": [
                        {
                          "type": "PrimConstructor",
                          "data": "String"
                        },
                        {
                          "type": "Call",
                          "data": {
                            "function": {
                              "type": "PrimConstructor",
                              "data": "Array"
                            },
                            "arguments": [
                              {
                                "type": "Constructor",
                                "data": {
                                  "constructor_kind": "Type",
                                  "canonical_value": {
                                    "module_name": [
                                      null,
                                      [
                                        "Test"
                                      ]
                                    ],
                                    "value": "Attr"
                                  },
                                  "source_value": {
                                    "module_name": null,
                                    "value": "Attr"
                                  }
                                }
                              }
                            ]
                          }
                        },
                        {
                          "type": "Call",
                          "data": {
                            "function": {
                              "type": "PrimConstructor",
                              "data": "Array"
                            },
                            "arguments": [
                              {
                                "type": "Call",
                                "data": {
                                  "function": {
                                    "type": "Constructor",
                                    "data": {
                                      "constructor_kind": {
                                        "Function": {
                                          "parameters": [
                                            {
                                              "Variable": 1
                                            }
                                          ]
                                        }
                                      },
                                      "canonical_value": {
                                        "module_name": [
                                          null,
                                          [
                                            "Test"
                                          ]
                                        ],
                                        "value": "Html"
                                      },
                                      "source_value": {
                                        "module_name": null,
                                        "value": "Html"
                                      }
                                    }
                                  },
                                  "arguments": [
                                    {
                                      "type": "Variable",
                                      "data": {
                                        "variable_kind": {
                                          "Variable": 1
                                        },
                                        "var": 0,
                                        "source_name": "msg"
                                      }
                                    }
                                  ]
                                }
                              }
                            ]
                          }
                        }
                      ],
                      "return_type": {
                        "type": "Call",
                        "data": {
                          "function": {
                            "type": "Constructor",
                            "data": {
                              "constructor_kind": {
                                "Function": {
                                  "parameters": [
                                    {
                                      "Variable": 1
                                    }
                                  ]
                                }
                              },
                              "canonical_value": {
                                "module_name": [
                                  null,
                                  [
                                    "Test"
                                  ]
                                ],
                                "value": "Html"
                              },
                              "source_value": {
                                "module_name": null,
                                "value": "Html"
                              }
                            }
                          },
                          "arguments": [
                            {
                              "type": "Variable",
                              "data": {
                                "variable_kind": {
                                  "Variable": 1
                                },
                                "var": 0,
                                "source_name": "msg"
                              }
                            }
                          ]
                        }
                      }
                    }
                  },
                  "variable": "h"
                }
              },
              "arguments": [
                {
                  "Expression": {
                    "expression": "String",
                    "data": {
                      "span": {
                        "start_offset": 165,
                        "end_offset": 170
                      },
                      "value": "div",
                      "value_type": {
                        "type": "PrimConstructor",
                        "data": "String"
                      }
                    }
                  }
                },
                {
                  "Expression": {
                    "expression": "LocalVariable",
                    "data": {
                      "span": {
                        "start_offset": 172,
                        "end_offset": 177
                      },
                      "variable_type": {
                        "type": "Call",
                        "data": {
                          "function": {
                            "type": "PrimConstructor",
                            "data": "Array"
                          },
                          "arguments": [
                            {
                              "type": "Constructor",
                              "data": {
                                "constructor_kind": "Type",
                                "canonical_value": {
                                  "module_name": [
                                    null,
                                    [
                                      "Test"
                                    ]
                                  ],
                                  "value": "Attr"
                                },
                                "source_value": {
                                  "module_name": null,
                                  "value": "Attr"
                                }
                              }
                            }
                          ]
                        }
                      },
                      "variable": "attrs"
                    }
                  }
                },
                {
                  "Expression": {
                    "expression": "LocalVariable",
                    "data": {
                      "span": {
                        "start_offset": 179,
                        "end_offset": 187
                      },
                      "variable_type": {
                        "type": "Call",
                        "data": {
                          "function": {
                            "type": "PrimConstructor",
                            "data": "Array"
                          },
                          "arguments": [
                            {
                              "type": "Call",
                              "data": {
                                "function": {
                                  "type": "Constructor",
                                  "data": {
                                    "constructor_kind": {
                                      "Function": {
                                        "parameters": [
                                          {
                                            "Variable": 1
                                          }
                                        ]
                                      }
                                    },
                                    "canonical_value": {
                                      "module_name": [
                                        null,
                                        [
                                          "Test"
                                        ]
                                      ],
                                      "value": "Html"
                                    },
                                    "source_value": {
                                      "module_name": null,
                                      "value": "Html"
                                    }
                                  }
                                },
                                "arguments": [
                                  {
                                    "type": "Variable",
                                    "data": {
                                      "variable_kind": {
                                        "Variable": 1
                                      },
                                      "var": 0,
                                      "source_name": "msg"
                                    }
                                  }
                                ]
                              }
                            }
                          ]
                        }
                      },
                      "variable": "children"
                    }
                  }
                }
              ]
            }
          }
        }
      }
    }
  },
  "values_toposort": [
    "div"
  ]
}